### PR TITLE
expression: rewrite builtin function: DIV

### PR DESCRIPTION
**Pull request by lkk2003rty**
to #3851

### CODE REVIEW

File: expression/builtin_arithmetic.go [hunk 1]
```diff
@@ -334,6 +340,107 @@ func (s *builtinArithmeticMinusIntSig) evalInt(row []types.Datum) (val int64, is
...
+		}
+		setFlenDecimal4RealOrDecimal(bf.tp, tpA, tpB, false)
+		sig := &builtinArithmeticDivideDecimalSig{baseDecimalBuiltinFunc{bf}}
+		return sig.setSelf(sig), nil
+	} else {
+		bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpDecimal, tpInt, tpInt)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		setFlenDecimal4Int(bf.tp, tpA, tpB)
```
Comment (zz-jason) [inline expression/builtin_arithmetic.go]:
```
drop table if exists t;
create table t(a bigint, b bigint);
insert into t select 1, 3;
select a/b from t;
```

result is:

```sql
mysql> select a/b from t;select a/b from t;
Field   1:  `a/b`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     25
Max_length: 6
Decimals:   4
Flags:      BINARY NUM


+--------+
| a/b    |
+--------+
| 0.3333 |
+--------+
1 row in set (0.00 sec)
```

seems we can not use `setFlenDecimal4Int` to set `Flen` and `Decimal` for result type of function `/`

File: expression/builtin_arithmetic.go [hunk 2]
```diff
@@ -334,6 +340,107 @@ func (s *builtinArithmeticMinusIntSig) evalInt(row []types.Datum) (val int64, is
...
+	sc := s.ctx.GetSessionVars().StmtCtx
+	a, isNull, err := s.args[0].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	b, isNull, err := s.args[1].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	if math.Abs(b) < floatEpsilon {
```
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: you can check the result of `a/b` is `nan` or `inf` and handle overflow error.

File: expression/builtin_arithmetic.go [hunk 3]
```diff
@@ -334,6 +340,107 @@ func (s *builtinArithmeticMinusIntSig) evalInt(row []types.Datum) (val int64, is
...
+		}
+		setFlenDecimal4Int(bf.tp, tpA, tpB)
+		sig := &builtinArithmeticDivideIntSig{baseDecimalBuiltinFunc{bf}}
+		return sig.setSelf(sig), nil
+	}
+}
+
+type builtinArithmeticDivideRealSig struct{ baseRealBuiltinFunc }
+type builtinArithmeticDivideDecimalSig struct{ baseDecimalBuiltinFunc }
+type builtinArithmeticDivideIntSig struct{ baseDecimalBuiltinFunc }
```
Comment (hanfei1991) [inline expression/builtin_arithmetic.go]: Is there any difference between divideInt and devideDecimal ?
Do we really need divideInt ?

File: expression/builtin_arithmetic.go [hunk 4]
```diff
@@ -334,6 +340,107 @@ func (s *builtinArithmeticMinusIntSig) evalInt(row []types.Datum) (val int64, is
...
+	}
+	b, isNull, err := s.args[1].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	if math.Abs(b) < floatEpsilon {
+		return 0, true, nil
+	}
+
+	return a / b, false, nil
```
Comment (hanfei1991) [inline expression/builtin_arithmetic.go]: check overflow:
mysql> select 1e300/1e-20;
ERROR 1690 (22003): DOUBLE value is out of range in '(1e300 / 1e-20)'

**Comment by zz-jason** [comment]
to https://github.com/pingcap/tidb/issues/4080

**Comment by hanfei1991** [comment]
dup with #4180

**Pull request closed without merging**